src/Lean/Compiler/IR/CompilerM.lean

```diff
@@ -82,13 +84,73 @@ private abbrev findAtSorted? (decls : Array Decl) (declName : Name) : Option Dec
   let tmpDecl := Decl.extern declName #[] default default
   decls.binSearch tmpDecl declLt
 
+namespace CollectUsedFDecls
+
+abbrev M := StateM NameSet
+
+@[inline] def collect (f : FunId) : M Unit :=
+  modify fun s => s.insert f
+
+partial def collectFnBody : FnBody → M Unit
+  | .vdecl _ _ v b   =>
+    match v with
+    | .fap f _ => collect f *> collectFnBody b
+    | .pap f _ => collect f *> collectFnBody b
+    | _        => collectFnBody b
+  | .jdecl _ _ v b   => collectFnBody v *> collectFnBody b
+  | .case _ _ _ alts => alts.forM fun alt => collectFnBody alt.body
+  | e => unless e.isTerminal do collectFnBody e.body
+
+def collectDecl : Decl → M NameSet
+  | .fdecl (body := b) .. => collectFnBody b *> get
+  | .extern .. => get
+
+end CollectUsedFDecls
+
+/-- Adds to `used` all `Decl.fdecl`s referenced directly by `decl`. -/
+def collectUsedFDecls (decl : Decl) (used : NameSet := {}) : NameSet :=
+  (CollectUsedFDecls.collectDecl decl).run' used
+
+/-- Computes the closure of `Decl.fdecl`s referenced by `decl`. -/
+def getFDeclClosure (m : DeclMap) (decls : Array Decl) : NameSet := Id.run do
+  let mut toVisit := decls.map (·.name) |>.toList
```

Comment:
Wouldn't this be better as an `Array`, changing the worklist algorithm to pop/push off of the back instead? Unfortunately, I think doing this requires a `repeat` instead of a `while`, because this isn't valid syntax:
```
  while let some n := toVisit? do
```

Reply:
It probably doesn't matter either way, does it? This list shouldn't be much more than a dozen entries long